src/hotspot/share/opto/superword.cpp

```diff
@@ -2820,111 +2871,130 @@ void VTransform::adjust_pre_loop_limit_to_align_main_loop_vectors() {
+         iv_scale  != 0 && is_power_of_2(abs(iv_scale))  &&
+         abs(iv_scale) < aw, "otherwise we cannot affect alignment with pre-loop");
 
-  const int AW = aw / abs(scale);
+  const int AW = aw / abs(iv_scale);
 
 #ifdef ASSERT
   if (_trace._align_vector) {
-    tty->print_cr("  AW = aw(%d) / abs(scale(%d)) = %d", aw, scale, AW);
+    tty->print_cr("  AW = aw(%d) / abs(iv_scale(%d)) = %d", aw, iv_scale, AW);
   }
 #endif
 
   // 1: Compute (13a, b):
-  //    xboi = -boi = (-base - offset - invar)         (stride * scale > 0)
-  //    xboi = +boi = (+base + offset + invar)         (stride * scale < 0)
-  const bool is_sub = scale * stride > 0;
-
-  // 1.1: offset
-  Node* xboi = igvn().intcon(is_sub ? -offset : offset);
-  TRACE_ALIGN_VECTOR_NODE(xboi);
-
-  // 1.2: invar (if it exists)
-  if (invar != nullptr) {
-    if (igvn().type(invar)->isa_long()) {
+  //    xbic = -bic = (-base - invar - con)         (iv_stride * iv_scale > 0)
+  //    xbic = +bic = (+base + invar + con)         (iv_stride * iv_scale < 0)
+  const bool is_sub = iv_scale * iv_stride > 0;
+
+  // 1.1: con
+  Node* xbic = igvn().intcon(is_sub ? -con : con);
+  TRACE_ALIGN_VECTOR_NODE(xbic);
+
+  // 1.2: invar = SUM(invar_summands)
+  //      We iteratively add / subtract all invar_summands, if there are any.
+  p.for_each_invar_summand([&] (const MemPointerSummand& s) {
+    Node* invar_variable = s.variable();
+    jint  invar_scale    = s.scale().value();
+    if (igvn().type(invar_variable)->isa_long()) {
       // Computations are done % (vector width/element size) so it's
       // safe to simply convert invar to an int and loose the upper 32
       // bit half.
-      invar = new ConvL2INode(invar);
-      phase()->register_new_node(invar, pre_ctrl);
-      TRACE_ALIGN_VECTOR_NODE(invar);
-   }
+      invar_variable = new ConvL2INode(invar_variable);
+      phase()->register_new_node(invar_variable, pre_ctrl);
+      TRACE_ALIGN_VECTOR_NODE(invar_variable);
+    }
+    Node* invar_scale_con = igvn().intcon(invar_scale);
+    Node* invar_summand = new MulINode(invar_variable, invar_scale_con);
+    phase()->register_new_node(invar_summand, pre_ctrl);
+    TRACE_ALIGN_VECTOR_NODE(invar_summand);
     if (is_sub) {
-      xboi = new SubINode(xboi, invar);
+      xbic = new SubINode(xbic, invar_summand);
     } else {
-      xboi = new AddINode(xboi, invar);
+      xbic = new AddINode(xbic, invar_summand);
     }
-    phase()->register_new_node(xboi, pre_ctrl);
-    TRACE_ALIGN_VECTOR_NODE(xboi);
-  }
+    phase()->register_new_node(xbic, pre_ctrl);
+    TRACE_ALIGN_VECTOR_NODE(xbic);
+  });
 
   // 1.3: base (unless base is guaranteed aw aligned)
-  if (aw > ObjectAlignmentInBytes || align_to_ref_p.base()->is_top()) {
-    // The base is only aligned with ObjectAlignmentInBytes with arrays.
-    // When the base() is top, we have no alignment guarantee at all.
-    // Hence, we must now take the base into account for the calculation.
```

Comment:
Note: `align_to_ref_p.base()->is_top()` meant it was a native memory access, i.e. off-heap, and not related to any java object.